torch/_dynamo/variables/torch.py

```diff
@@ -1096,32 +1005,40 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
             graph_checkpoint, checkpoint = tx.output.graph, tx.copy_graphstate()
 
             def speculate_branch(branch):
+                # NB: 0 is predicate
+                ix = 1 if branch else 2
                 try:
-                    # NB: 0 is predicate
-                    ix = 1 if branch else 2
-                    return old_speculate_subgraph(
-                        args[ix], operands, graph_checkpoint, checkpoint
+                    ret_val, ret_graph, ret_lifted_freevars = speculate_subgraph(
+                        tx, args[ix], operands, graph_checkpoint, checkpoint
                     )
-                except ArgsMismatchError as e:
-                    raise UserError(UserErrorType.DYNAMIC_CONTROL_FLOW, str(e))
+                # Reraise because we want to suggest workarounds
+                except Unsupported as e:
+                    raise UserError(UserErrorType.DYNAMIC_CONTROL_FLOW, str(e)) from e
+
+                if not isinstance(ret_val, TensorVariable):
+                    raise UserError(
+                        UserErrorType.DYNAMIC_CONTROL_FLOW,
+                        "Expected branch out type to be a single tensor",
+                    )
+                return ret_val, ret_graph, ret_lifted_freevars
 
-            (
-                true_r,
-                true_graph,
-                true_guards,
-                true_nn_modules_context,
-                true_cmp,
-            ) = speculate_branch(True)
-            (
-                false_r,
-                false_graph,
-                false_guards,
-                false_nn_modules_context,
-                false_cmp,
-            ) = speculate_branch(False)
+            (true_r, true_graph, true_lifted_freevars) = speculate_branch(True)
+
+            state_after_true_branch = tx.copy_graphstate()
+            true_guards = state_after_true_branch.output.guards
+            true_nn_modules = state_after_true_branch.output.nn_modules
+            true_cmp = get_comparable_state(state_after_true_branch)
+
+            (false_r, false_graph, false_lifted_freevars) = speculate_branch(False)
+
+            state_after_false_branch = tx.copy_graphstate()
+            false_guards = state_after_false_branch.output.guards
+            false_nn_modules = state_after_false_branch.output.nn_modules
+            false_cmp = get_comparable_state(state_after_false_branch)
```

Comment:
A lot of this logic is redundant and we should just delete most of it. `speculate_subgraph` already (by design) mutates tx.output's state (guards, tracked_fakes, tracking), so there should be no need for us to add additioanl things to it.

This function should just look like:
```
(true_r, true_graph, true_lifted_freevars) = speculate_branch(True)
(false_r, false_graph, false_lifted_freevars) = speculate_branch(False)
fixup_branch_inps(true_branch)
fixup_branch_inps(false_branch)
add_subgraph(true_branch)
add_subgraph(false_branch)
```
If you agree, then this also should apply to the map() implementation below.

Also, I'm not really sure how to test that we've propagated the guards/fake_tensor values/etc correctly. We should probably add a test that we're not regressing that

Reply:
Yep seems to be working after I delete all these.